projects/pgai/pgai/semantic_catalog/vectorizer/vectorizer.py

```diff
@@ -0,0 +1,214 @@
+from collections.abc import Sequence
+from typing import Annotated, Any, Literal
+
+import psycopg
+from psycopg.rows import dict_row
+from psycopg.sql import SQL, Identifier
+from pydantic import BaseModel, Field
+
+
+class SentenceTransformersConfig(BaseModel):
+    implementation: Literal["sentence_transformers"]
+    config_type: Literal["embedding"]
+    model: str
+    dimensions: int
+
+
+class OllamaConfig(BaseModel):
+    implementation: Literal["ollama"]
+    config_type: Literal["embedding"]
+    model: str
+    dimensions: int
+    base_url: str | None = None
+
+
+class OpenAIConfig(BaseModel):
+    implementation: Literal["openai"]
+    config_type: Literal["embedding"]
+    model: str
+    dimensions: int
+    base_url: str | None = None
+    api_key_name: str | None = None
+
+
+EmbeddingConfig = Annotated[
+    SentenceTransformersConfig | OllamaConfig | OpenAIConfig,
+    Field(discriminator="implementation"),
+]
+
+
+def embedding_config_from_dict(config: dict[str, Any]) -> EmbeddingConfig:
+    assert "implementation" in config
+    match config["implementation"]:
+        case "sentence_transformers":
+            return SentenceTransformersConfig(**config)
+        case "ollama":
+            return OllamaConfig(**config)
+        case "openai":
+            return OpenAIConfig(**config)
+        case _:
+            raise ValueError(
+                f"Unrecognized embedding implementation: {config['implementation']}"
+            )
+
+
+class EmbedRow(BaseModel):
+    id: int
+    content: str
+    vector: Sequence[float] | None = None
+
+
+async def _get_obj_batch(
+    con: psycopg.AsyncConnection, id: int, name: str, batch_size: int
+) -> list[EmbedRow]:
+    async with con.cursor(row_factory=dict_row) as cur:
+        await cur.execute(
+            SQL("""\
+            select
+              id
+            , concat_ws
+              ( ' '
+              , case
+                when objargs is not null and array_length(objargs, 1) > 0 then
+                    format
+                    ( '%%s(%%s)'
+                    , array_to_string(objnames, '.')
+                    , array_to_string(objargs, ', ')
+                    )
+                else array_to_string(objnames, '.')
+                end
+              , description
+              ) as content
+            from ai.{table}
+            where {name} is null
```

Comment:
do we need some randomness in the row orderings?

Reply:
is there a better way than just `order by random()`?

Reply:
random is the way

Reply:
[6ed5a35](https://github.com/timescale/pgai/pull/617/commits/6ed5a35b563b2cd308bc71391a38c7e890262f8c)